lib/AST/Decl.cpp

```diff
@@ -8333,6 +8334,33 @@ AbstractFunctionDecl::getDerivativeFunctionConfigurations() {
     ctx.loadDerivativeFunctionConfigurations(this, previousGeneration,
                                              *DerivativeFunctionConfigs);
   }
+
+  class DerivativeFinder : public ASTWalker {
```

Comment:
Drive-by question: does `@derivative` attribute finding have an impact on incremental compilation?

---

When a file is updated, it and all files that depend on it need to be recompiled. Incremental compilation is fast when there are few dependencies between files.

I wonder if `@derivative` finding can cause incremental compilation to become asymptotically worse. Thinking through an example:

1. `a.swift` defines `func foo`. `b.swift` exists in the same module, defining a `func vjpFoo` function with no attributes.
2. Many files in the same module ask for the derivative of `foo`, e.g. by calling `gradient(of: foo)`.

Initially, `foo` has no derivative functions registered via `@derivative`, so `gradient(of: foo)` triggers the compiler to automatically generate derivative functions.

3. `func vjpFoo` in`b.swift` is now marked with `@derivative(of: foo)`.

Since all files (`a.swift`, `b.swift`, `gradient(of: foo)` files) are in the same module, the addition of `@derivative(of: foo)` in `b.swift` means that `foo` now has a registered derivative, and all functions that call `gradient(of: foo)` need to be recompiled†.

Is the † recompilation behavior asymptotically worse than existing incremental compilation? It might not be, if the same amount of work (e.g. recompiling the entire module) was necessary previous to the implementation of correct `@derivative` attribute finding.

Reply:
(It may be that the impact of `@derivative` attribute finding on incremental compilation was introduced prior to this PR.)

Reply:
> The complete recompilation is certainly necessary when optimizations are enabled.

When optimizations are enabled you're building in whole-module mode anyways. Swift will refuse to perform the incremental build if you ask for -wmo as well.

An even simpler failure mode occurs here since this analysis is not going through name lookup. From the incremental build's perspective, there is no link between the derivative and the original.

Reply:
> When optimizations are enabled you're building in whole-module mode anyways.

I'm confused. Will you please elaborate? The original testcase from https://github.com/apple/swift/issues/55170 fails regardless whether `-O` is provided to `swiftc`.